Better handling of Autograd+Fork errors.

VitalyFedyunin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#33885 Better handing of Autograd+Fork errors.**

Fixes: #32835
Fixes: #5834

Can not combine with CUDA's implementation as each of them requires individual `std::once_flag` as well as different `forked_autograd_child` functions. CUDA version relays to python module, autograd uses TORCH_CHECK to report error to python and cpp.

Differential Revision: [D20144024](https://our.internmc.facebook.com/intern/diff/D20144024)

VitalyFedyunin committed:
Better handing of Autograd+Fork errors.

[ghstack-poisoned]

albanD commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -728,16 +746,24 @@ auto Engine::execute(const edge_list& roots,
+void Engine::initialize_threads_pool() {
+  track_bad_autograd_forks();
+  TORCH_CHECK(!in_bad_autograd_fork,
+              "Unable to handle autograd's threading in combination with fork. "
```
nit: "fork" -> "fork-based multiprocessing"
Maybe some people use the multiprocessing package without knowing what fork is?

albanD reviewed (approved):
LGTM

VitalyFedyunin committed:
Update on "Better handing of Autograd+Fork errors."


Fixes: #32835
Fixes: #5834

Can not combine with CUDA's implementation as each of them requires individual `std::once_flag` as well as different `forked_autograd_child` functions. CUDA version relays to python module, autograd uses TORCH_CHECK to report error to python and cpp.



[ghstack-poisoned]

VitalyFedyunin committed:
Update on "Better handing of Autograd+Fork errors."


Fixes: #32835
Fixes: #5834

Can not combine with CUDA's implementation as each of them requires individual `std::once_flag` as well as different `forked_autograd_child` functions. CUDA version relays to python module, autograd uses TORCH_CHECK to report error to python and cpp.



[ghstack-poisoned]

VitalyFedyunin committed:
Update on "Better handing of Autograd+Fork errors."


Fixes: #32835
Fixes: #5834

Can not combine with CUDA's implementation as each of them requires individual `std::once_flag` as well as different `forked_autograd_child` functions. CUDA version relays to python module, autograd uses TORCH_CHECK to report error to python and cpp.

Differential Revision: [D20144024](https://our.internmc.facebook.com/intern/diff/D20144024)

[ghstack-poisoned]

albanD commented on test/test_multiprocessing.py:
```diff
@@ -821,15 +834,6 @@ def test_is_shared_cuda(self):
         t = torch.randn(5, 5).cuda()
         self.assertTrue(t.is_shared())
 
-    @unittest.skip('this test occasionally fails and deadlocks; see https://github.com/pytorch/pytorch/issues/5834')
```
Nice catch !
